src/messages/permissions/CommandPermissions.ts

```diff
@@ -10,89 +11,110 @@ import Common from '../../common';
 import DiscordClient from '../../DiscordClient';
 import Settings, { SettingsJSON } from '../../Settings';
 import { ApplicationCommandType } from '../MessageHandler';
-import Authentication from './Authentication';
+import Authentication, { MaybeTokenResponse } from './Authentication';
 import { RequestOptions } from 'https';
+import GuildSettings from '../../settings/GuildSettings';
 
 /**
  * Set permissions for commands
  */
 export default class CommandPermissions {
   #commands?: Collection<string, ApplicationCommandType>;
+  #guild: Guild;
+
+  constructor(guild: Guild) {
+    this.#guild = guild;
+  }
 
   /**
    * Is to be called after commands have been set
    * @param commands The commands to set permissions for
    */
-  onCommandsSet(commands: Collection<Snowflake, ApplicationCommandType>): void {
+  async onCommandsSet(
+    commands: Collection<Snowflake, ApplicationCommandType>
+  ): Promise<void> {
     this.#commands = commands;
-    if (process.env.DISCORD_REFRESH_TOKEN !== undefined) {
-      if (!Common.accessTokenValid())
-        Authentication.getAccessToken()
+
+    const refreshToken = await Authentication.getRefreshToken(this.#guild.id);
+
+    if (refreshToken !== undefined) {
+      if (!Common.accessTokenValid(this.#guild.id))
+        Authentication.getAccessToken(this.#guild)
           .then(this.setPermissions.bind(this))
           .catch((err: Error): void => {
             if (err.message !== 'invalid_grant') throw new Error(err.message);
-            new Authentication(this.setPermissions.bind(this));
+
+            Authentication.startServer();
+            Authentication.addListener(
+              this.#guild.id,
+              this.setPermissions.bind(this)
+            );
           });
       else this.setPermissions();
       return;
     }
-    if (Settings.settings['logging'] !== 'errors')
-      new Authentication(this.setPermissions.bind(this));
+    Authentication.startServer();
+    Authentication.addListener(this.#guild.id, this.setPermissions.bind(this));
   }
 
   /**
    * Set permissions for commands, is used as callback for when the user has been authorized and an access token is available
    */
-  private async setPermissions(): Promise<void> {
+  private async setPermissions(json?: MaybeTokenResponse): Promise<void> {
+    Authentication.removeListener(this.#guild.id);
+    if (json !== undefined) await Authentication.storeToken(json, this.#guild);
```

Comment:
This is a side effect of the function. Maybe move to somewhere else?

Reply:
This is actually not a side effect but rather the function is poorly called and does too much. Instead it should be renamed to something like `accessTokenReceived` and the permission set should be extracted into a different function this function calls.